File: 2025-03-29/Group_Anagrams.java
package amazon;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;

public class Group_Anagrams {
    public static List<List<String>> groupAnagrams(String[] strs) {
        if(strs == null || strs.length == 0) return new ArrayList<>();

        HashMap<String, ArrayList<String>> map = new HashMap<>();

        for(String s:strs) {
            char[] arr = s.toCharArray();
            Arrays.sort(arr);
            String sorted = new String(arr);

            map.computeIfAbsent(sorted, k->new ArrayList<>()).add(s);
        }

        return new ArrayList<>(map.values());
    }
    public static void main(String[] args) {
        String[] str = {"eat","tea","tan","ate","nat","bat"};
        System.out.println(groupAnagrams(str));
    }
}

File: 2026-01-16/Group_Anagrams.py
from collections import defaultdict
from typing import List


class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        if not strs:
            return []

        anagram_map = defaultdict(list)

        # brute force - o(n * m * logm)
        # for str in strs:
        #     sorted_str = "".join(sorted(str))
        #     anagram_map[sorted_str] .append(str)

        # optimized(26 chars) - o(n * m)
        # for str in strs:
        #     freq = [0] * 26
        #     for c in str:
        #         freq[ord(c) - ord('a')] += 1
        #
        #     anagram_map[tuple(freq)].append(str)

        # Optimal(all chars with unicode) - o(n * (m + ulogu))
        for str in strs:
            freq = defaultdict(int)
            for c in str:
                freq[c] += 1

            key = tuple(sorted(freq.items()))
            anagram_map[key].append(str)

        return list(anagram_map.values())


sol = Solution()
print(sol.groupAnagrams(["eat","tea","tan","ate","nat","bat"]))